tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+	uint64 stat_input_valid_rows;
+	uint64 stat_bulk_filtered_rows;
+} GroupingPolicyHash;
+
+static const GroupingPolicy grouping_policy_hash_functions;
+
+GroupingPolicy *
+create_grouping_policy_hash(List *agg_defs, List *output_grouping_columns)
+{
+	GroupingPolicyHash *policy = palloc0(sizeof(GroupingPolicyHash));
+	policy->funcs = grouping_policy_hash_functions;
+	policy->output_grouping_columns = output_grouping_columns;
+	policy->agg_defs = agg_defs;
+	policy->agg_extra_mctx =
+		AllocSetContextCreate(CurrentMemoryContext, "agg extra", ALLOCSET_DEFAULT_SIZES);
+	policy->allocated_aggstate_rows = TARGET_COMPRESSED_BATCH_SIZE;
+	ListCell *lc;
+	foreach (lc, agg_defs)
+	{
+		VectorAggDef *agg_def = lfirst(lc);
+		policy->per_agg_states =
+			lappend(policy->per_agg_states,
+					palloc0(agg_def->func.state_bytes * policy->allocated_aggstate_rows));
+	}
+
+	policy->table = h_create(CurrentMemoryContext, policy->allocated_aggstate_rows, NULL);
+	policy->have_null_key = false;
+
+	policy->returning_results = false;
+
+	return &policy->funcs;
+}
+
+static void
+gp_hash_reset(GroupingPolicy *obj)
+{
+	GroupingPolicyHash *policy = (GroupingPolicyHash *) obj;
+
+	MemoryContextReset(policy->agg_extra_mctx);
+
+	policy->returning_results = false;
+
+	h_reset(policy->table);
+	policy->have_null_key = false;
+
+	policy->stat_input_valid_rows = 0;
+	policy->stat_input_total_rows = 0;
+	policy->stat_bulk_filtered_rows = 0;
+}
+
+static void
+compute_single_aggregate(DecompressBatchState *batch_state, int start_row, int end_row,
+						 VectorAggDef *agg_def, void *agg_states, uint32 *offsets,
+						 MemoryContext agg_extra_mctx)
+{
+	ArrowArray *arg_arrow = NULL;
+	Datum arg_datum = 0;
+	bool arg_isnull = true;
+
+	/*
+	 * We have functions with one argument, and one function with no arguments
+	 * (count(*)). Collect the arguments.
+	 */
+	if (agg_def->input_offset >= 0)
+	{
+		CompressedColumnValues *values = &batch_state->compressed_columns[agg_def->input_offset];
+		Assert(values->decompression_type != DT_Invalid);
+		Assert(values->decompression_type != DT_Iterator);
+
+		if (values->arrow != NULL)
+		{
+			arg_arrow = values->arrow;
+		}
+		else
+		{
+			Assert(values->decompression_type == DT_Scalar);
+			arg_datum = *values->output_value;
+			arg_isnull = *values->output_isnull;
+		}
+	}
+
+	/*
+	 * Now call the function.
+	 */
+	if (arg_arrow != NULL)
+	{
+		/* Arrow argument. */
+		agg_def->func
+			.agg_many_vector(agg_states, offsets, start_row, end_row, arg_arrow, agg_extra_mctx);
+	}
+	else
+	{
+		/*
+		 * Scalar argument, or count(*). The latter has an optimized
+		 * implementation.
+		 */
+		if (agg_def->func.agg_many_scalar != NULL)
+		{
+			agg_def->func.agg_many_scalar(agg_states,
+										  offsets,
+										  start_row,
+										  end_row,
+										  arg_datum,
+										  arg_isnull,
+										  agg_extra_mctx);
+		}
+		else
+		{
+			for (int i = start_row; i < end_row; i++)
+			{
+				if (offsets[i] == 0)
```

Comment:
Isn't 0 a legitimate offset into the agg_states array? Or is it 1-indexed?

Reply:
I'm using 0 as invalid, added some comments about that.